File: leetcode-go/Q001-Q100/Q047_Permutations_II/Q047_Permutations_II.go
package Q047_Permutations_II


import (
//	"sort"
	"fmt"
	"sort"
)

func permuteUnique(nums []int) [][]int {
	sort.Ints(nums)
	visit := make([]bool, len(nums))
	ans := [][]int{}
	gao(&ans, []int{}, nums, visit)
	return ans
}

func gao(ans *[][]int, perms []int, nums []int, visit []bool) {
	if len(perms) == len(nums) {
		*ans = append(*ans, perms)
		return
	}

	for i := 0; i < len(nums); i++ {
		if visit[i] {
			continue
		}
		if i > 0 && nums[i-1] == nums[i] && !visit[i-1] {
			continue
		}
		visit[i] = true
		//perms =
		gao(ans, append(perms, nums[i]), nums, visit)
		visit[i] = false
		//perms = perms[:len(perms)-1]
		/*
		for i+1 < len(nums) && nums[i+1] == nums[i] {
			i++
		}
		*/
	}
}

func main() {
	permuteUnique([]int{1, 2, 3, 4})
}

File: src/Q001_Q300/Q047_Permutations_II.java
package Q001_Q300;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by mqliang on 6/4/17.
 */
public class Q047_Permutations_II {
    public class Solution {
        public List<List<Integer>> permuteUnique(int[] nums) {
            Arrays.sort(nums);
            List<List<Integer>> ans = new ArrayList<>();
            boolean[] visit = new boolean[nums.length];
            dfs(ans, visit, new ArrayList<>(), nums);
            return ans;
        }

        public void dfs(List<List<Integer>> ans, boolean[] visit, List<Integer> combs, int[] nums) {
            if (combs.size() == nums.length) {
                ans.add(new ArrayList<>(combs));
                return;
            }
            for (int i = 0; i < nums.length; i++) {
                if (visit[i]) continue;
                visit[i] = true;
                combs.add(nums[i]);
                dfs(ans, visit, combs, nums);
                combs.remove(combs.size() - 1);
                visit[i] = false;
                while (i + 1 < nums.length && nums[i + 1] == nums[i]) i++;
            }
        }
    }
}
